### ISSUE TITLE
Graph optimization results in broken model

### ISSUE THREAD

**Issue by david-macleod** [microsoft/onnxruntime#20951]
### Describe the issue

When using `ORT_ENABLE_EXTENDED` on an onnxruntime session the graph it produces is broken, specifically it uses the incorrect tensor as the input to an operation. This does not happen with optimizations disabled.

The following shows a min-repro.

This is the un-optimized graph, note how the input `x` is used to derive the `shape` that is an input to `add`. This is correct.
<img width="410" alt="image" src="https://github.com/microsoft/onnxruntime/assets/17232877/876734e8-62cd-4e9c-b968-e7385f87a4a4">

After optimization the other input `state.in` is now uses to derive the `shape` that is an input to `add`, which results in an incorrect result.

<img width="481" alt="image" src="https://github.com/microsoft/onnxruntime/assets/17232877/20ca1a5b-fe31-4210-a783-e23e1814b655">

This only happens on the nightly version of onnxruntime, and not the previous major release.

### To reproduce

```python
import torch
import onnxruntime as ort

class Model(torch.nn.Module):

    def __init__(self):
        super().__init__()

    def forward(self, x, state):

        x = x + x.shape[1]

        if state.shape[0] == 1:
            state = state + 1
        y = torch.cat([state, x], dim=1)

        return y, state


model = torch.jit.script(Model())

output_path = "model.onnx"
input_names = ["x", "state.in"]
output_names = ["y", "state.out"]
dynamic_axes = {"x": {0: "a", 1: "b"}, "state.in": {0: "a", 1: "b"},
                "y": {0: "a", 1: "b"},"state.out": {0: "a", 1: "b"}}

torch.onnx.export(model, (torch.randn(1,4), torch.randn(1,0)), output_path,
                  input_names=input_names, output_names=output_names, dynamic_axes=dynamic_axes)


options = ort.SessionOptions()
options.graph_optimization_level = ort.GraphOptimizationLevel.ORT_ENABLE_EXTENDED
options.optimized_model_filepath = output_path.replace(".onnx", "_optimized.onnx")

session = ort.InferenceSession(output_path, options, providers=["CPUExecutionProvider"])
```

### Urgency

Medium

### Platform

Linux

### OS Version

22.04

### ONNX Runtime Installation

Released Package

### ONNX Runtime Version or Commit ID

ort_nightly-1.19.0.dev20240603001-cp310-cp310-manylinux_2_27_x86_64.manylinux_2_28_x86_64.whl

### ONNX Runtime API

Python

### Architecture

X64

### Execution Provider

Default CPU

**Comment by yuslepukhin** [comment]
Extended level really runs only one optimizer AFAIK.
You can selectively disable optimizers by name in python API and see which one is the problem.

**Comment by david-macleod** [comment]
I still get the same behaviour with `ort.GraphOptimizationLevel.ORT_ENABLE_BASIC`, is that what you mean?

**Comment by mindest** [comment]
This behavior should not be considered as an error I think, since graph inputs `x` and `state.in` have the same (symbolic) shape.
The change was introduced by #19832, with which the graph transformer only keeps one `Shape` node if there are multiple with the same shape value.

**Comment by david-macleod** [comment]
In what way do `x` and `state.in` have the same shape? They have the same number of dimensions but the axes are dynamic and can take on any value (as per the example inputs in the export code).

**Comment by mindest** [comment]
> In what way do `x` and `state.in` have the same shape? They have the same number of dimensions but the axes are dynamic and can take on any value (as per the example inputs in the export code).

You set `dynamic_axes` for `x` and `state.in` (`a` for dim 0 and `b` for dim 1), which resulted in the same symbolic shapes for them. If that is not what you want (like the example inputs with shapes (1, 4) and (1, 0)), you can set them with different symbolic shapes, like:

```diff
- dynamic_axes = {"x": {0: "a", 1: "b"}, "state.in": {0: "a", 1: "b"},
-                 "y": {0: "a", 1: "b"},"state.out": {0: "a", 1: "b"}}
+ dynamic_axes = {"x": {0: "a", 1: "b1"}, "state.in": {0: "a", 1: "b2"}}
```

and let it infer the output shapes. Will this meet your need?

**Comment by david-macleod** [comment]
Ok great, thanks! Have made sure dimensions that are different have unique names and all working now
